# HR Employee Attrition & Workforce Analysis

## Overview
This project is a full-cycle data analyst workflow using a real-world Human Resources dataset. The goal is to simulate the kind of work a junior data analyst does day-to-day: cleaning messy raw data, exploring it, answering business questions across multiple tools, and building a final dashboard for stakeholders.

## Dataset
The dataset contains employee-level HR records including demographic information, employment history, compensation, performance, and termination details. Key fields include:

- **Identifiers**: Employee_Name, EmpID, ManagerID, DeptID, PositionID
- **Demographics**: DOB, Sex, MaritalDesc, RaceDesc, CitizenDesc, HispanicLatino
- **Employment**: DateofHire, DateofTermination, EmploymentStatus, Termd, TermReason, Department, Position, ManagerName
- **Compensation & Performance**: Salary, PerformanceScore, EngagementSurvey, EmpSatisfaction, SpecialProjectsCount
- **Recruitment**: RecruitmentSource, FromDiversityJobFair
- **Attendance**: DaysLateLast30, Absences
- **Location**: State, Zip

The raw data contains realistic inconsistencies including missing values, redundant/overlapping columns describing the same attribute, inconsistent categorical labels, and mixed date formats.

## Objective
To analyze workforce composition, performance, and attrition patterns in order to answer key business questions such as:

- Which departments have the highest attrition rates, and why
- Is there a relationship between employee engagement/satisfaction and attrition
- How does performance score correlate with salary and tenure
- Which recruitment sources produce the most retained, highest-performing employees
- Are there patterns in absenteeism or lateness tied to attrition risk

## Tools and Workflow
This project is structured across four stages, each using a different tool to mirror the range of skills expected in a data analyst role:

1. **Python (Pandas)** – Data cleaning and exploratory data analysis. Handling missing values, resolving inconsistent categorical entries, standardizing date formats, reconciling redundant columns, and preparing a clean dataset for analysis.
2. **Excel** – Business question analysis using pivot tables, formulas, and summary views.
3. **SQL** – Querying the cleaned dataset to answer business questions using joins, aggregations, subqueries, window functions, and CTEs.
4. **Power BI** – Building an interactive dashboard to visualize workforce and attrition insights for stakeholders.

## Deliverables
- Cleaned dataset (CSV)
- Jupyter Notebook documenting the cleaning and EDA process
- Excel workbook with business question analysis
- SQL script with queries and explanations
- Power BI dashboard (PBIX file and/or published report)

In [7]:
import pandas as pd 

In [8]:
df = pd.read_csv('HR_dataset.csv')

In [9]:
df.head(5)

,Employee_Name,EmpID,MarriedID,MaritalStatusID,GenderID,EmpStatusID,DeptID,PerfScoreID,FromDiversityJobFairID,Salary,...,ManagerName,ManagerID,RecruitmentSource,PerformanceScore,EngagementSurvey,EmpSatisfaction,SpecialProjectsCount,LastPerformanceReview_Date,DaysLateLast30,Absences
0,"Adinolfi, Wilson K",10026,0,0,1,1,5,4,0,62506,...,Michael Albert,22.0,LinkedIn,Exceeds,4.60,5,0,1/17/2019,0,1
1,"Ait Sidi, Karthikeyan",10084,1,1,1,5,3,3,0,104437,...,Simon Roup,4.0,Indeed,Fully Meets,4.96,3,6,2/24/2016,0,17
2,"Akinkuolie, Sarah",10196,1,1,0,5,5,3,0,64955,...,Kissy Sullivan,20.0,LinkedIn,Fully Meets,3.02,3,0,5/15/2012,0,3
3,"Alagbe,Trina",10088,1,1,0,1,5,3,0,64991,...,Elijiah Gray,16.0,Indeed,Fully Meets,4.84,5,0,1/3/2019,0,15
4,"Anderson, Carol",10069,0,2,0,5,5,3,0,50825,...,Webster Butler,39.0,Google Search,Fully Meets,5.00,4,0,2/1/2016,0,2


In [15]:
print(f'rows: {df.shape[0]} and columns: {df.shape[1]}')

rows: 311 and columns: 36


In [16]:
df.dtypes

Employee_Name                  object
EmpID                           int64
MarriedID                       int64
MaritalStatusID                 int64
GenderID                        int64
EmpStatusID                     int64
DeptID                          int64
PerfScoreID                     int64
FromDiversityJobFairID          int64
Salary                          int64
Termd                           int64
PositionID                      int64
Position                       object
State                          object
Zip                             int64
DOB                            object
Sex                            object
MaritalDesc                    object
CitizenDesc                    object
HispanicLatino                 object
RaceDesc                       object
DateofHire                     object
DateofTermination              object
TermReason                     object
EmploymentStatus               object
Department                     object
ManagerName 

In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 311 entries, 0 to 310
Data columns (total 36 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   Employee_Name               311 non-null    object 
 1   EmpID                       311 non-null    int64  
 2   MarriedID                   311 non-null    int64  
 3   MaritalStatusID             311 non-null    int64  
 4   GenderID                    311 non-null    int64  
 5   EmpStatusID                 311 non-null    int64  
 6   DeptID                      311 non-null    int64  
 7   PerfScoreID                 311 non-null    int64  
 8   FromDiversityJobFairID      311 non-null    int64  
 9   Salary                      311 non-null    int64  
 10  Termd                       311 non-null    int64  
 11  PositionID                  311 non-null    int64  
 12  Position                    311 non-null    object 
 13  State                       311 non

In [13]:
df.isnull().sum()

Employee_Name                   0
EmpID                           0
MarriedID                       0
MaritalStatusID                 0
GenderID                        0
EmpStatusID                     0
DeptID                          0
PerfScoreID                     0
FromDiversityJobFairID          0
Salary                          0
Termd                           0
PositionID                      0
Position                        0
State                           0
Zip                             0
DOB                             0
Sex                             0
MaritalDesc                     0
CitizenDesc                     0
HispanicLatino                  0
RaceDesc                        0
DateofHire                      0
DateofTermination             207
TermReason                      0
EmploymentStatus                0
Department                      0
ManagerName                     0
ManagerID                       8
RecruitmentSource               0
PerformanceSco

In [18]:
df.duplicated().sum()

np.int64(0)

In [19]:
df['EmpID'].nunique()

311

In [20]:
df.nunique()

Employee_Name                 311
EmpID                         311
MarriedID                       2
MaritalStatusID                 5
GenderID                        2
EmpStatusID                     5
DeptID                          6
PerfScoreID                     4
FromDiversityJobFairID          2
Salary                        308
Termd                           2
PositionID                     30
Position                       32
State                          28
Zip                           158
DOB                           307
Sex                             2
MaritalDesc                     5
CitizenDesc                     3
HispanicLatino                  4
RaceDesc                        6
DateofHire                    101
DateofTermination              96
TermReason                     18
EmploymentStatus                3
Department                      6
ManagerName                    21
ManagerID                      23
RecruitmentSource               9
PerformanceSco

In [22]:
categorical_cols = ['Sex', 'MaritalDesc', 'CitizenDesc', 'HispanicLatino', 
                     'RaceDesc', 'TermReason', 'EmploymentStatus', 
                     'Department', 'RecruitmentSource', 'PerformanceScore',
                     'Position', 'State']

for col in categorical_cols:
    print(f"\n{col}:")
    print(df[col].unique())


Sex:
['M ' 'F']

MaritalDesc:
['Single' 'Married' 'Divorced' 'Widowed' 'Separated']

CitizenDesc:
['US Citizen' 'Eligible NonCitizen' 'Non-Citizen']

HispanicLatino:
['No' 'Yes' 'no' 'yes']

RaceDesc:
['White' 'Black or African American' 'Two or more races' 'Asian'
 'American Indian or Alaska Native' 'Hispanic']

TermReason:
['N/A-StillEmployed' 'career change' 'hours' 'return to school'
 'Another position' 'unhappy' 'attendance' 'performance'
 'Learned that he is a gangster' 'retiring' 'relocation out of area'
 'more money' 'military' 'no-call, no-show' 'Fatal attraction'
 'maternity leave - did not return' 'medical issues' 'gross misconduct']

EmploymentStatus:
['Active' 'Voluntarily Terminated' 'Terminated for Cause']

Department:
['Production       ' 'IT/IS' 'Software Engineering' 'Admin Offices'
 'Sales' 'Executive Office']

RecruitmentSource:
['LinkedIn' 'Indeed' 'Google Search' 'Employee Referral'
 'Diversity Job Fair' 'On-line Web application' 'CareerBuilder' 'Website'
 'Other

# Data Cleaning Plan

## Objective
Before moving into exploratory analysis and business questions, the raw HR dataset needs to be cleaned. Initial exploration using `.info()`, `.describe()`, `.isnull().sum()`, and `.nunique()` revealed inconsistencies typical of real-world HR data: whitespace issues, inconsistent text casing, and redundant columns that need to be cross-validated against each other.

## Issues Identified

### 1. Whitespace inconsistencies
- `Sex` — values like `'M '` contain trailing whitespace
- `Department` — values like `'Production       '` contain excessive trailing whitespace
- `Position` — `'Data Analyst'` and `'Data Analyst '` exist as separate values due to trailing whitespace

### 2. Inconsistent text casing
- `HispanicLatino` — contains `'Yes'`, `'No'`, `'yes'`, `'no'` as four separate values instead of two
- `TermReason` — mixed casing across entries (e.g. `'career change'` vs `'Another position'`)

### 3. Data quality anomalies
- `TermReason` contains two entries that appear to be placeholder or joke text (`'Learned that he is a gangster'`, `'Fatal attraction'`) rather than genuine termination reasons. These need to be recoded (e.g. to `'Other'`) rather than deleted, to avoid losing legitimate rows over a labeling issue.

### 4. Column redundancy requiring validation
Three columns describe the same underlying fact — whether an employee is still active or has left the company:
- `Termd` (0/1 flag)
- `EmploymentStatus` (Active / Voluntarily Terminated / Terminated for Cause)
- `TermReason` (text reason, or `'N/A-StillEmployed'` if active)

Since these columns should logically agree with each other, a crosstab check is used to confirm there are no contradictions (e.g. an employee marked `Termd = 0` but `EmploymentStatus = 'Voluntarily Terminated'`).

## Planned Actions
1. Strip whitespace from `Sex`, `Department`, and `Position`
2. Standardize casing on `HispanicLatino` to two consistent values
3. Standardize casing on `TermReason` and recode placeholder entries to `'Other'`
4. Run `pd.crosstab(df['EmploymentStatus'], df['Termd'])` to confirm status and termination flag agree
5. Run a check confirming `TermReason` correctly shows `'N/A-StillEmployed'` for every `Active` employee

In [23]:
df['Sex'] = df['Sex'].str.strip()

In [24]:
df['HispanicLatino'] = df['HispanicLatino'].str.strip().str.capitalize()

In [25]:
df['Department'] = df['Department'].str.strip()

In [27]:
df['Position'] = df['Position'].str.strip()
df['Position'].nunique()

31

In [28]:
df['TermReason'] = df['TermReason'].str.strip().str.lower()
df['TermReason'] = df['TermReason'].replace({
    'learned that he is a gangster': 'other',
    'fatal attraction': 'other'
})

In [29]:
pd.crosstab(df['EmploymentStatus'], df['Termd'])

Termd,0,1
EmploymentStatus,,
Active,207,0
Terminated for Cause,0,16
Voluntarily Terminated,0,88


In [30]:
pd.crosstab(df['EmploymentStatus'], df['TermReason'] == 'N/A-StillEmployed')

TermReason,False
EmploymentStatus,
Active,207
Terminated for Cause,16
Voluntarily Terminated,88


In [31]:
df['TermReason'].value_counts()

TermReason
n/a-stillemployed                   207
another position                     20
unhappy                              14
more money                           11
career change                         9
hours                                 8
attendance                            7
return to school                      5
relocation out of area                5
performance                           4
retiring                              4
military                              4
no-call, no-show                      4
medical issues                        3
maternity leave - did not return      3
other                                 2
gross misconduct                      1
Name: count, dtype: int64

# Cleaning Conclusion: Status Consistency Check

## What Was Checked
A crosstab was run between `EmploymentStatus` and `Termd` to confirm the two columns agree on employment status across all 311 records.

## Result

| EmploymentStatus         | Termd = 0 | Termd = 1 |
|---------------------------|-----------|-----------|
| Active                     | 207       | 0         |
| Terminated for Cause       | 0         | 16        |
| Voluntarily Terminated     | 0         | 88        |

The result is a clean diagonal with no contradictions. Every employee marked `Active` has `Termd = 0`, and every terminated employee (for cause or voluntary) has `Termd = 1`. This confirms `Termd` and `EmploymentStatus` are fully consistent with each other, and no rows need correction on this basis.

## Note on TermReason Check
A second check compared `EmploymentStatus` against whether `TermReason` equals `'N/A-StillEmployed'`. The result requires re-verification, as the row counts (207 / 16 / 88) matched the `EmploymentStatus` breakdown exactly but all returned `False`, which is inconsistent with the expectation that all 207 `Active` employees should show `'N/A-StillEmployed'` as `True`. This should be re-run using `df['TermReason'].value_counts()` to directly confirm the count of `'N/A-StillEmployed'` entries before finalizing this check.

## Outcome
`Termd` and `EmploymentStatus` are confirmed consistent. No data correction needed for this pair. The `TermReason` validation is pending re-verification before this column can be considered fully checked.

In [32]:
date_cols = ['DOB', 'DateofHire', 'DateofTermination', 'LastPerformanceReview_Date']

for col in date_cols:
    df[col] = pd.to_datetime(df[col], errors='coerce')

C:\Users\user\AppData\Local\Temp\ipykernel_13236\2096369669.py:4: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  df[col] = pd.to_datetime(df[col], errors='coerce')


In [33]:
for col in date_cols:
    print(col, df[col].isnull().sum())

DOB 0
DateofHire 0
DateofTermination 207
LastPerformanceReview_Date 0


In [34]:
df[df['ManagerID'].isnull()][['Employee_Name', 'Position', 'ManagerName', 'Department']]

,Employee_Name,Position,ManagerName,Department
19,"Becker, Scott",Production Technician I,Webster Butler,Production
30,"Buccheri, Joseph",Production Technician II,Webster Butler,Production
44,"Chang, Donovan E",Production Technician I,Webster Butler,Production
88,"Fancett, Nicole",Production Technician II,Webster Butler,Production
135,"Hutter, Rosalie",Production Technician II,Webster Butler,Production
177,"Manchester, Robyn",Production Technician II,Webster Butler,Production
232,"Rivera, Haley",Production Technician I,Webster Butler,Production
251,"Sewkumar, Nori",Production Technician I,Webster Butler,Production


In [42]:
# df[df['ManagerName'] == 'Webster Butler']['ManagerID'].unique() [39.nan] -> Before
df['ManagerID'] = df['ManagerID'].fillna(
    df.loc[df['ManagerName'] == 'Webster Butler', 'ManagerID'].mode()[0]
)
# df[df['ManagerName'] == 'Webster Butler']['ManagerID'].unique() [39.] -> After
df['ManagerID'] = df['ManagerID'].fillna(39)
df['ManagerID'].isnull().sum()

np.int64(0)